gensim/models/ldamodel.py

```diff
@@ -83,6 +83,87 @@ def logsumexp(x):
 
     return x
 
+def mean_absolute_difference(a, b):
+    """
+    Mean absolute difference between two arrays
+
+    Parameters
+    ----------
+    a : (M,) array_like of float32
+    b : (M,) array_like of float32
+
+    Returns
+    -------
+    float32
+        mean(abs(a - b))
+    
+    """
+    return np.mean(np.abs(a - b))
+
+try:
+    # try to load fast, cythonized code if possible
+    from gensim.models.ldamodel_inner import dirichlet_expectation_1d_f32, dirichlet_expectation_1d_f64
+    from gensim.models.ldamodel_inner import dirichlet_expectation_2d_f32, dirichlet_expectation_2d_f64
+    from gensim.models.ldamodel_inner import logsumexp_2d_f32, logsumexp_2d_f64
+    from gensim.models.ldamodel_inner import mean_absolute_difference_f32, mean_absolute_difference_f64
+    FAST_VERSION = 1
+except ImportError:
+    # else fall back to python/numpy
+    FAST_VERSION = -1
+    
+    logsumexp_2d_f32 = logsumexp_2d_f64 = logsumexp
+    dirichlet_expectation_1d_f32 = dirichlet_expectation_1d_f64 = dirichlet_expectation
+    dirichlet_expectation_2d_f32 = dirichlet_expectation_2d_f64 = dirichlet_expectation
+    mean_absolute_difference_f32 = mean_absolute_difference_f64 = mean_absolute_difference
+
+
+def _attach_inner_methods(self, dtype):
```

Comment:
Probably better to define needed functions on module level (not "patch" `LdaModel` at runtime), this allows avoiding a problem with backward compatibility.

Reply:
that was my original intention (and that's what the original "fastldamodel.py" code did), but I am not sure how to make this work nicely with LdaModel since a user could create multiple models with different dtypes.  Eg,

model0 = LdaModel(corpus=corpus, id2word=corpus.id2word, num_topics=10, dtype=np.float16)
model1 = LdaModel(corpus=corpus, id2word=corpus.id2word, num_topics=10, dtype=np.float32)
model2 = LdaModel(corpus=corpus, id2word=corpus.id2word, num_topics=10, dtype=np.float64)

each of those objects would need to use different implementations of the inner functions (unless we dispatched at run-time based on dtypes, but as mentioned before, i think that will have a decent impact on speed).

Reply:
Agree, you depend on `dtype` of LDA

Reply:
Aha, you are right, `dtype` is a really fresh feature (for LDA), I forgot about it.
Ok, probably that's variant is OK (another variant - add this as method to LDA and call in init)

Reply:
I believe we should decide what dtype makes the most sense, and stick with it (drop support for all other dtype).

It's not a parameter we can ask users to choose for themselves (especially since we give no guidance, no pros/cons/recommendations).

Reply:
@piskvorky probably `float32` should be the best choice (by speed, memory consumption and quality)

Reply:
I agree, but would like to see some numeric arguments (benchmarks, model quality & performance evals) to back that up.

Reply:
removing the other dtypes would certainly make this change easier. 

I was running some tests with the different dtypes, and float16 (one of the allowed options in the existing code) seems to over/underflow with some reasonable input data.  For example the following code (I left out the imports) will crash with an overflow error:

np.seterr(all='raise')

text = api.load('text8')
id2word = gensim.corpora.Dictionary(text)
corpus = [ id2word.doc2bow(t) for t in text ]

LdaModel(corpus=corpus, id2word=id2word, num_topics=10, dtype=np.float16)